File: 000401-000600/000441.排列硬币/441.js
/**
 * @param {number} n
 * @return {number}
 */

 // 方法二: 用时 88ms
 // 执行用时超过了 87%
 // 消耗内存超过了 18%
var arrangeCoins = function(n) {
  let left = 1
  let right = n
  while (left < right) {
    let mid = Math.floor((right - left + 1) / 2) + left
    if (mid * (mid + 1) <= 2 * n) {
      left = mid
    } else {
      right = mid - 1
    }
  }
  return left
};

console.log(arrangeCoins(5))
console.log(arrangeCoins(8))

File: 000401-000600/000441.排列硬币/441.go
package main

import (
	"fmt"
)

// 方法二: 用时 4ms
// 执行用时超过了 73%
// 消耗内存超过了 99%
func arrangeCoins(n int) int {
	left := 0
	right := n
	for ;left < right; {
		mid := int((right - left + 1) / 2) + left
		if (mid + 1) * mid <= 2 * n {
			left = mid
		} else {
			right = mid - 1
		}
	}
	return left
}

func main ()  {
	fmt.Println(arrangeCoins(5))
	fmt.Println(arrangeCoins(8))
}
